File: algorithm/challenges/11.vowels/main.js
// Basic solution
function vowels1(str) {
  let count = 0;
  for (let char of str.toLowerCase()) {
    if (
      char === "a" ||
      char === "e" ||
      char === "i" ||
      char === "o" ||
      char === "u"
    )
      count++;
  }
  return count;
}

// With includes
function vowels2(str) {
  let count = 0;
  for (let char of str.toLowerCase()) {
    if ("aeiou".includes(char)) count++;
  }
  return count;
}

// With regex
function vowels3(str) {
  const volwelsArray = str.match(/[aeiou]/gi); // if matches are found, an array of matches is returned. If no match is found, return null
  const n = volwelsArray ? volwelsArray.length : 0;
  return n;
}

// Testing
const test = require('node:test');
const assert = require('node:assert');

test('returns the number of vowels used', () => {
  assert.strictEqual(vowels1('aeiou'), 5);
  assert.strictEqual(vowels1('AEIOU'), 5);
  assert.strictEqual(vowels1('abcdefghijklmnopqrstuvwxyz'), 5);
  assert.strictEqual(vowels1('bcdfghjkl'), 0);
});

File: algorithm/challenges/11.vowels/main.clj
(require '[clojure.test :as test])

(defn number-of-vowels
  [string]
  (->> string
       (re-seq #"[aeiou]")
       count))

(test/deftest number-of-vowels-test
  (test/testing "Check the number of vowels in a string"
    (test/is (= 5 (number-of-vowels "abcdefghijklmnopqrstuvwxyz")))
    (test/is (= 0 (number-of-vowels "bcdfghjkl")))))

(test/run-test number-of-vowels-test)
